R/getMainfolder.R

```diff
@@ -1,61 +1,68 @@
 #' getMainfolder
-#' 
+#'
 #' Functions checks for a global setting of the mainfolder (either by setting the environment
 #' variable "MADRAT_MAINFOLDER" or by setting the R option with the same name). If none of these
 #' is available the user will be asked for a directory. If this is not provided a temporary folder
-#' will be used. 
-#' 
+#' will be used.
+#'
 #' @param verbose boolean deciding whether status information/updates should be shown or not
 #' @param .testmode boolean switch only relevant for internal testing (will simulate user inputs)
 #' @author Jan Philipp Dietrich
 #' @seealso \code{\link{initializeConfig}}, \code{\link{getConfig}}, \code{\link{setConfig}}
-#' 
-getMainfolder <- function(verbose=TRUE, .testmode = FALSE) {
-  
+#'
+getMainfolder <- function(verbose = TRUE, .testmode = FALSE) {
   # option MADRAT_MAINFOLDER set?
   mainfolder <- getOption("MADRAT_MAINFOLDER")
-  if(!is.null(mainfolder)) return(mainfolder)
-  
+  if (!is.null(mainfolder)) return(mainfolder)
+
   # environment variable MADRAT_MAINFOLDER set?
   mainfolder <- Sys.getenv("MADRAT_MAINFOLDER", unset = NA)
-  if(!is.na(mainfolder)) return(mainfolder)
-  
-  # ask for folder to use
-  if(interactive() || .testmode){
-    repeat {
-      s <- tolower(readline("madrat mainfolder for data storage not set! Do you want to set it now? (y/n) "))
-      if(.testmode) s <- "y"
-      if(s %in% c("y","n")) break
+  if (!is.na(mainfolder)) return(mainfolder)
+
+  questionLoop <- function(question, testModeAnswer) {
+    if (.testmode) {
```

Comment:
What's the advantage of doing it that way over the way it's implemented now?

Reply:
R allows to directly access variables in parent environments, but this is dangerous as the function declaration creates the believe that the function only depends on `question` and `testModeAnswer` where in fact it also requires to have `.testmode` available. Therefor, I always try to avoid these situations and always try to put all variables required by a function into its list of arguments

Reply:
I agree that pure functions are the most robust kind of functions, but I strongly disagree with calling this dangerous. I'd say this is maybe a 2/10 danger score, where global variables aka options are at least a 7/10 and NSE is 9/10, and we are using that all over the place in madrat. In fact I would even argue that in this case an additional .testmode variable in the inner function is more confusing, because then you have two different variables with the same name which could have different values. I'd say this kind of shadowing is more dangerous than just using the outer .testmode in the inner function. Also, if I see an inner function I don't expect it to be independent of the parent scope, because one of the main reasons to use inner functions is precisely that they can use the outer variables. In lapply calls it is also very common to create inner functions which use variables from the enclosing scope.

Reply:
Are you okay with the new default argument solution I just pushed?